tt_metal/common/work_split.cpp

```diff
@@ -0,0 +1,234 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//
+// Contains utility functions for partitioning work between multiple cores.
+//
+
+#include <cstdint>
+#include <tuple>
+#include <vector>
+
+#include "tt_metal/common/assert.hpp"
+#include "tt_metal/common/core_coord.hpp"
+#include "tt_metal/common/math.hpp"
+#include "tt_metal/third_party/tracy/public/tracy/Tracy.hpp"
+
+namespace tt {
+namespace tt_metal {
+
+uint32_t merge_num_sticks_to_read(uint32_t num_sticks_to_read, uint32_t stick_size_bytes, uint32_t max_read_size) {
+    uint32_t total_bytes = num_sticks_to_read * stick_size_bytes;
+    uint32_t new_num_sticks_to_read = num_sticks_to_read;
+    uint32_t new_stick_size_bytes = stick_size_bytes;
+
+    for (uint32_t current_size = stick_size_bytes; current_size <= max_read_size; current_size += stick_size_bytes) {
+        if (total_bytes % current_size == 0) {
+            new_stick_size_bytes = current_size;
+            new_num_sticks_to_read = total_bytes / current_size;
+        }
+    }
+    return new_num_sticks_to_read;
+}
+
+// Given a number of tiles and number of cores available
+// Set the largest number of cores less than the number of tiles
+// Returns the number of cores as well as the number of tiles per core
+
+std::tuple<uint32_t, uint32_t> get_max_cores_divisible_by_tiles_per_core_tiles(
+    const uint32_t &num_tiles, const uint32_t &num_cores_max, bool request_even) {
+    uint32_t num_cores = 1;
+    for (int i = 2; i <= num_cores_max; i++) {
+        if ((num_tiles % i) == 0) {
+            num_cores = i;
+        }
+    }
+    if (request_even) {
+        num_cores = num_cores - num_cores % 2;
+    }
+    uint32_t per_core_tiles_dim = num_tiles / num_cores;
+    if (num_tiles % num_cores != 0)
+        per_core_tiles_dim++;
+    return {num_cores, per_core_tiles_dim};
+}
+
+// Finds the maximum even divisor of val starting at start_max_div and below
```

Comment:
I don't know how much I like having identical comments both in the hpp and cpp files. Maybe just have one?

Also, I know this is copied code, but what happens if you call e.g.`find_max_divisor(12,3)`, is it desired?
- If yes and the result is not even please update the comment description change.
- If no should it be fixed or would you open a new issue?

Reply:
I honestly don't know what the expected functionality of this should be. This was added a long time ago by someone to find a block/subblock sizes?, and I'm guessing odd numbers were either bugged or not supported when these settings were applied in kernels/llks, yet for some reason didn't check 3? For something generically called `find_max_divisor` then I don't think it should care about even/odd. If it did the function should be renamed. I'd rather not touch anything related to this in this pr, we can file an issue for it.

Reply:
Then it would be good to update the comment to replace `even` with something more accurate, and just have one comment instead of two.

Reply:
I've removed the comments from the cpp file, only in hpp now.